relayer/network_utils.go

```diff
@@ -29,116 +30,78 @@ func InitializeConnectionsAndCheckStake(
-					err,
-				)
-			}
-		}
+	for _, sourceBlockchainConfig := range cfg.SourceBlockchains {
+		network.TrackSubnet(sourceBlockchainConfig.GetSubnetID())
 	}
-	return nil
-}
+	ctx, cancel := context.WithTimeout(
+		context.Background(),
+		time.Duration(cfg.InitialConnectionTimeoutSeconds)*time.Second,
+	)
+	defer cancel()
 
-// Connect to the validators of the source blockchain. For each destination blockchain,
-// verify that we have connected to a threshold of stake.
-func connectToNonPrimaryNetworkPeers(
-	logger logging.Logger,
-	network peers.AppRequestNetwork,
-	cfg *config.Config,
-	sourceBlockchain *config.SourceBlockchain,
-) error {
-	subnetID := sourceBlockchain.GetSubnetID()
-	network.TrackSubnet(subnetID)
-	connectedValidators, err := network.GetConnectedCanonicalValidators(subnetID)
-	if err != nil {
-		logger.Error(
-			"Failed to connect to canonical validators",
-			zap.String("subnetID", subnetID.String()),
-			zap.Error(err),
-		)
-		return err
-	}
-	for _, destination := range sourceBlockchain.SupportedDestinations {
-		blockchainID := destination.GetBlockchainID()
-		if ok, warpConfig, err := checkForSufficientConnectedStake(logger, cfg, connectedValidators, blockchainID); !ok {
-			logger.Warn(
-				"Failed to connect to a threshold of stake",
-				zap.String("destinationBlockchainID", blockchainID.String()),
-				zap.Uint64("connectedWeight", connectedValidators.ConnectedWeight),
-				zap.Uint64("totalValidatorWeight", connectedValidators.ValidatorSet.TotalWeight),
-				zap.Any("WarpConfig", warpConfig),
-			)
-			return err
-		}
+	eg, ctx := errgroup.WithContext(ctx)
+	for _, sourceBlockchain := range cfg.SourceBlockchains {
+		eg.Go(func() error {
+			return checkSufficientConnectedStake(ctx, logger, network, cfg, sourceBlockchain)
+		})
 	}
-	return nil
+	return eg.Wait()
 }
 
-// Connect to the validators of the destination blockchains. Verify that we have connected
-// to a threshold of stake for each blockchain.
-func connectToPrimaryNetworkPeers(
+// Returns when connected to sufficient stake for all supported destinations of the source blockchain or
+// in case of errors or timeouts.
+func checkSufficientConnectedStake(
+	ctx context.Context,
 	logger logging.Logger,
 	network peers.AppRequestNetwork,
 	cfg *config.Config,
 	sourceBlockchain *config.SourceBlockchain,
 ) error {
+	subnetID := sourceBlockchain.GetSubnetID()
+	// Loop over destination blockchains here to confirm connections to a threshold of stake
+	// which is determined by the Warp Quorum configs of the destination blockchains.
 	for _, destination := range sourceBlockchain.SupportedDestinations {
-		blockchainID := destination.GetBlockchainID()
-		subnetID := cfg.GetSubnetID(blockchainID)
-		network.TrackSubnet(subnetID)
-		connectedValidators, err := network.GetConnectedCanonicalValidators(subnetID)
+		destinationBlockchainID := destination.GetBlockchainID()
+		warpConfig, err := cfg.GetWarpConfig(destinationBlockchainID)
```

Comment:
We should loop over all the `SupportedDestinations` to find the largest `quorumNumerator` and then only enter the below `for` loop once